### DISCUSSION TITLE
REQUEST: Ability to add/remove line item properties & reject discount codes in checkout

### DISCUSSION THREAD

**Discussion by rgilbert82** [Shopify/function-examples#413] [Feedback: Cart Transform API]
Hello. There are still a few useful things that were available in the Script Editor that have not yet been implemented in the new functions. Specifically:

1. The ability to add, edit or remove a line item's properties. Sometimes we use the existence or non-existence of line item properties to display different messaging in cart. We are currently using the Script Editor to change line item properties based on certain discount operations that have or have not occurred.

2. The ability to reject discount codes in checkout. For example, we might want to reject some or all discount codes depending on certain types of free gifts added to cart.

3. The ability to split line items. Line items with quantities greater than 1 might need to be split and treated differently. In some cases a line item might need to removed completely or its quantity might need to be reduced.

These are probably all good candidates for the Cart Transform API specifically. Thank you!

**Comment by SaeedYasin** [comment]
Here are some examples related to discount codes, which merchants are asking for,

1) Customer enters wrong discount but is new customer then show him x% off code
2) Customer enters wrong discount code. Show a custom error message depending on customer segmentation
3) Customer enters correct discount code. But applies to a different collection. Show info message to explain this fact

These are probably best suited for validation functions API.

**Comment by anthony-v-sundae-lane** [comment]
Additional use case for this is to add personalisation items to bundles.

ie.
Gift Pack product
- A customisable bundle that users can choose 3 items for + add some text for personalisation of a card or message.

Personalisations like a 'to'/'from' field are currently captured by line item properties within themes, but this is not currently compatible with bundle products.

**Comment by anthony-v-sundae-lane** [comment]
@ignacio-chiazzo - Just bumping this request to see if there's any update.

Our bundles app is now getting requests to have gift messages captured via line item properties for our mix and match bundles.
